File: backend/api/schemas/pose_sets.py
import uuid

from pydantic import BaseModel, Field, field_validator


class PoseSetCreateRequest(BaseModel):
    garment_id: uuid.UUID
    model_id: uuid.UUID
    cloth_type: str
    pose_ids: list[uuid.UUID] = Field(min_length=1)

    @field_validator("cloth_type")
    @classmethod
    def validate_cloth_type(cls, value: str) -> str:
        if value not in {"upper_body", "lower_body", "dress"}:
            raise ValueError("cloth_type must be upper_body, lower_body, or dress")
        return value


class PoseSetCreateResponse(BaseModel):
    pose_set_id: uuid.UUID
    batch_id: uuid.UUID
    total_items: int


class PoseSetItemResponse(BaseModel):
    pose_type: str | None
    batch_item_id: uuid.UUID
    media_id: uuid.UUID | None
    image_url: str | None
    status: str
    error_message: str | None


class PoseSetDetailResponse(BaseModel):
    pose_set_id: uuid.UUID
    batch_id: uuid.UUID
    garment_id: uuid.UUID
    model_id: uuid.UUID
    status: str
    items: list[PoseSetItemResponse]

File: frontend/lib/api/pose-sets.ts
import { apiFetch } from "@/lib/api";

export interface PoseSetCreateRequest {
  garment_id: string;
  model_id: string;
  cloth_type: "upper_body" | "lower_body" | "dress";
  pose_ids: string[];
}

export interface PoseSetCreateResponse {
  pose_set_id: string;
  batch_id: string;
  total_items: number;
}

export interface PoseSetItem {
  pose_type: "front" | "side" | "back" | null;
  batch_item_id: string;
  media_id: string | null;
  image_url: string | null;
  status: "pending" | "processing" | "complete" | "failed";
  error_message: string | null;
}

export interface PoseSetDetail {
  pose_set_id: string;
  batch_id: string;
  garment_id: string;
  model_id: string;
  status: string;
  items: PoseSetItem[];
}

export async function createPoseSet(
  request: PoseSetCreateRequest
): Promise<PoseSetCreateResponse> {
  return apiFetch("/api/pose-sets", {
    method: "POST",
    body: JSON.stringify(request),
  });
}

export async function getPoseSet(poseSetId: string): Promise<PoseSetDetail> {
  return apiFetch(`/api/pose-sets/${poseSetId}`);
}
